### PR TITLE
skip chunk if already in place

### PR DESCRIPTION
**Pull request by dgloeck**
I just noticed that I have reimplemented what you already started two years ago in your read-before-write branch. The commit calculates the hash of a chunk before it tries to fetch the chunk from a source.

What was missing in your branch is the handling of `last_id`. Since checking a target chunk overwrites the buffer, `have_last_id` must be updated. To not let the previous chunk go to waste, I skip checking the target and always write the chunk if `last_id` matches.

The code for adding a chunk to the index is moved to a separate function to be reused. I do not print any statistics about the skipped chunks.

### CODE REVIEW

**Comment by florolf** [comment]
Thank you for submitting this patch.

> I just noticed that I have reimplemented what you already started two years ago in your read-before-write branch.

It's been a while and I don't have access to the hardware/benchmark results anymore, but if I recall correctly, the reason I abandoned that approach was that it actually decreased the net performance. In general, the hit rate would not be very good because chunks tend to move around and you pay for this check with an additional read and hash operation for every chunk. The main place where this would help is preventing store queries for long runs of all-zeroes chunks, which is already (asymptotically) handled through caching the last chunk we wrote.

On the storage side I think it could help when you have a large enough read/write performance asymmetry to tip the balance or your FTL doesn't handle rewriting identical contents into a block properly and wears out the storage medium.

In the end, it didn't really work out on the platform I tried it on, but this might be different depending on the layout/filesystem of the image you're syncing and the hardware you're running on. I'd be curious about your setup and benchmark results if you have any.

About the patch itself:

Looks fine to me, but the `last_id` logic in `entry_cb` is getting a bit hard to follow. `have_last_id` signals that `last_id` has been initialized/matches whatever `buf` contains. Maybe `target_check_chunk` could have an out parameter for the ID it calculated for the contents of `tmp`. Then we could point it to `last_id` and a call to `target_check_chunk` would maintain this invariant unless it errored out because of a failed `preadall`.

> Since checking a target chunk overwrites the buffer, have_last_id must be updated. To not let the previous chunk go to waste, I skip checking the target and always write the chunk if last_id matches.

Depending on why this patch actually helps in your application, this probably wastes some potential for dropping writes, e.g. with runs of all-zeroes chunks in the input, which are probably also already present on the target.

I think adding a second buffer for the target read would be acceptable. Or, considering that the whole `last_id` logic empirically only fires for all-zeroes chunks anyway, we could just do away with it entirely and hardcode this as a special case, leaving the buffer space free for a check like the one you implemented.

**Comment by dgloeck** [comment]
I have changed the commit to use a second buffer if the first buffer already holds the correct chunk data. I have kept your buffer alignment of 4096 although I doubt anything above cache line size will make a difference. Maybe if there is a crypto accelerator the kernel might be able to use one dma descriptor less during hashing...

Our motivation for this PR is not performance. We have devices with unstable internet connection and without this patch casync-nano will start downloading chunks from the beginning every time it has to be restarted.

**Comment by florolf** [comment]
Sorry for letting this sit around for so long.

I agree that not being able to resume updates really is a glaring omission now that you have pointed it out, thanks!

Still, I feel that the arguments above about the cost/benefit ratio of performing this check all the time still holds. However, if the main thing you're concerned about is resuming interrupted updates, we can just do that check until we encounter the first mismatch and get most of the benefits with little additional cost. I've implemented this [here](https://github.com/florolf/casync-nano/commit/d5b7fe7cbfd782e985125adb1e9c75f8ad36613d). What do you think?

>  I have kept your buffer alignment of 4096 although I doubt anything above cache line size will make a difference. Maybe if there is a crypto accelerator the kernel might be able to use one dma descriptor less during hashing...

It's been a long time but I think a number of drivers dropped into a (very) slow path when the buffer wasn't page-aligned.

**Comment by dgloeck** [comment]
> I've implemented this [here](https://github.com/florolf/casync-nano/commit/d5b7fe7cbfd782e985125adb1e9c75f8ad36613d). What do you think?

Looks like a good compromise!

**Comment by florolf** [comment]
Included in v0.4
